core/services/gateway/handlers/functions/allowlist/allowlist_test.go

```diff
@@ -0,0 +1,184 @@
+	}, testutils.WaitTimeout(t), time.Second).Should(gomega.BeTrue())
+}
+func TestAllowlist_UpdateFromContract(t *testing.T) {
+	t.Parallel()
+
+	t.Run("OK-iterate_over_list_of_allowed_senders", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(testutils.Context(t))
+		client := mocks.NewClient(t)
+		client.On("LatestBlockHeight", mock.Anything).Return(big.NewInt(42), nil)
+		client.On("CallContract", mock.Anything, mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+			cancel()
+		}).Return(sampleEncodedAllowlist(t), nil)
+		config := allowlist.OnchainAllowlistConfig{
+			ContractAddress:            common.HexToAddress(addr3),
+			ContractVersion:            1,
+			BlockConfirmations:         1,
+			UpdateFrequencySec:         1,
+			UpdateTimeoutSec:           1,
+			StoredAllowlistBatchSize:   2,
+			OnchainAllowlistBatchSize:  16,
+			StoreAllowedSendersEnabled: true,
+			FetchingDelayInRangeSec:    0,
+		}
+
+		orm := amocks.NewORM(t)
+		orm.On("DeleteAllowedSenders", []common.Address{common.HexToAddress(addr1), common.HexToAddress(addr2)}).Times(2).Return(nil)
+		orm.On("CreateAllowedSenders", []common.Address{common.HexToAddress(addr1), common.HexToAddress(addr2)}).Times(2).Return(nil)
+
+		allowlist, err := allowlist.NewOnchainAllowlist(client, config, orm, logger.TestLogger(t))
+		require.NoError(t, err)
+
+		err = allowlist.UpdateFromContract(ctx)
+		require.NoError(t, err)
+
+		gomega.NewGomegaWithT(t).Eventually(func() bool {
+			return allowlist.Allow(common.HexToAddress(addr1)) && !allowlist.Allow(common.HexToAddress(addr3))
+		}, testutils.WaitTimeout(t), time.Second).Should(gomega.BeTrue())
+	})
+
+	t.Run("OK-fetch_complete_list_of_allowed_senders_without_storing", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(testutils.Context(t))
+		client := mocks.NewClient(t)
+		client.On("LatestBlockHeight", mock.Anything).Return(big.NewInt(42), nil)
+		client.On("CallContract", mock.Anything, mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+			cancel()
+		}).Return(sampleEncodedAllowlist(t), nil)
+		config := allowlist.OnchainAllowlistConfig{
+			ContractAddress:            common.HexToAddress(addr3),
+			ContractVersion:            1,
+			BlockConfirmations:         1,
+			UpdateFrequencySec:         1,
+			UpdateTimeoutSec:           1,
+			StoredAllowlistBatchSize:   2,
+			OnchainAllowlistBatchSize:  16,
+			StoreAllowedSendersEnabled: false,
+			FetchingDelayInRangeSec:    0,
+		}
+
+		orm := amocks.NewORM(t)
+		allowlist, err := allowlist.NewOnchainAllowlist(client, config, orm, logger.TestLogger(t))
+		require.NoError(t, err)
+
+		err = allowlist.UpdateFromContract(ctx)
+		require.NoError(t, err)
+
+		gomega.NewGomegaWithT(t).Eventually(func() bool {
+			return allowlist.Allow(common.HexToAddress(addr1)) && !allowlist.Allow(common.HexToAddress(addr3))
+		}, testutils.WaitTimeout(t), time.Second).Should(gomega.BeTrue())
+	})
```

Comment:
We should probably add a test to check that blocked senders are properly removed from the db.

Reply:
Im checking this on the `"OK-iterate_over_list_of_allowed_senders"` test by checking the ORM mock is called:

```
orm.On("DeleteAllowedSenders", []common.Address{common.HexToAddress(addr1), common.HexToAddress(addr2)}).Times(2).Return(nil)
orm.On("CreateAllowedSenders", []common.Address{common.HexToAddress(addr1), common.HexToAddress(addr2)}).Times(2).Return(nil)
```

but its true it would be better having a specific test to do this, the problem I found is that we are mocking the contract on every call with the same payload:
```
client.On("CallContract", mock.Anything, mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
			cancel()
		}).Return(sampleEncodedAllowlist(t), nil)
....
}
		
func sampleEncodedAllowlist(t *testing.T) []byte {
	abiEncodedAddresses :=
		"0000000000000000000000000000000000000000000000000000000000000020" +
			"0000000000000000000000000000000000000000000000000000000000000002" +
			"000000000000000000000000" + addr1 +
			"000000000000000000000000" + addr2
	rawData, err := hex.DecodeString(abiEncodedAddresses)
	require.NoError(t, err)
	return rawData
}
```

Haven't found a way to mock specific underlying contract methods, do you know how to do this?

Reply:
I think you should be able to match based on the specific contract function that is being called instead of using `mock.Anything`.
You are more than welcomed to refactor the tests to do this, but I think just mocking the ORM call is sufficient given that is in line with the testing we previously had and we have separate unit tests which cover the ORM.